staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go

```diff
@@ -21,39 +21,254 @@ import (
 	switch target.(type) {
 	case *corev1.Pod:
 		// do nothing to the copied pod
 		return nil
 	case *corev1.Node:
-		const volumeName = "host-root"
-		pod.Spec.Volumes = append(pod.Spec.Volumes, corev1.Volume{
-			Name: volumeName,
-			VolumeSource: corev1.VolumeSource{
-				HostPath: &corev1.HostPathVolumeSource{Path: "/"},
-			},
+		MountRootPartition(pod, containerName)
+		UseHostNamespaces(pod)
+		return nil
+	default:
+		return fmt.Errorf("the %s profile doesn't support objects of type %T", ProfileLegacy, target)
+	}
+}
+
+type debugStyle int
+
+const (
+	// debug by ephemeral container
+	styleEphemeral debugStyle = iota
+	// debug by pod copy
+	stylePodCopy
+	// debug node
+	styleNode
+)
+
+func getDebugStyle(pod *corev1.Pod, target runtime.Object) (debugStyle, error) {
+	switch target.(type) {
+	case *corev1.Pod:
+		if asserted, ok := target.(*corev1.Pod); ok {
+			if pod != asserted { // comparing addresses
+				return stylePodCopy, nil
+			}
+		}
+		return styleEphemeral, nil
+	case *corev1.Node:
+		return styleNode, nil
+	}
+	return debugStyle(-1), fmt.Errorf("objects of type %T are not supported", target)
+}
+
+func (p *generalProfile) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
+	style, err := getDebugStyle(pod, target)
+	if err != nil {
+		return fmt.Errorf("general profile: %s", err)
+	}
+
+	switch style {
+	case styleNode:
+		MountRootPartition(pod, containerName)
+		ClearSecurityContext(pod, containerName, Containers)
+		UseHostNamespaces(pod)
+
+	case stylePodCopy:
+		RemoveLabelsAndProbes(pod)
+		AllowProcessTracing(pod, containerName, Containers)
+		ShareProcessNamespace(pod)
+
+	case styleEphemeral:
+		AllowProcessTracing(pod, containerName, EphemeralContainers)
+	}
+
+	return nil
+}
+
+func (p *baselineProfile) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
+	style, err := getDebugStyle(pod, target)
+	if err != nil {
+		return fmt.Errorf("baseline profile: %s", err)
+	}
+
+	ClearSecurityContext(pod, containerName, Containers|EphemeralContainers)
+
+	switch style {
+	case stylePodCopy:
+		RemoveLabelsAndProbes(pod)
+		ShareProcessNamespace(pod)
+
+	case styleEphemeral, styleNode:
+		// no additional modifications needed
+	}
+
+	return nil
+}
+
+func (p *restrictedProfile) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
+	style, err := getDebugStyle(pod, target)
+	if err != nil {
+		return fmt.Errorf("restricted profile: %s", err)
+	}
+
+	DisallowRoot(pod, containerName, Containers|EphemeralContainers)
+	DropCapabilities(pod, containerName, Containers|EphemeralContainers)
+
+	switch style {
+	case styleNode:
+		ClearSecurityContext(pod, containerName, Containers)
+
+	case stylePodCopy:
+		ShareProcessNamespace(pod)
+
+	case styleEphemeral:
+		// no additional modifications needed
+	}
+
+	return nil
+}
+
+// ContainerType signifies container type
+type ContainerType int
```

Comment:
We have already `debugStyle`, I think there is no need additional type field?

Reply:
It would be nice to use one concept with both, but I'm afraid that `ContainerType` and `debugStyle` might grow in ways that we may not want to mix the two concepts.

For example, a future debug profile might want to make modifications to init containers when debugging by pod copy. So it needs to add a new `ContainerType` of `InitContainer`. Conversely, a future debugging methodology might target a new runtime object type that is not a container type, which would require adding a new value of `debugStyle`.